File: lib/features/admin/application/use_cases/select_customer_profile_photo.dart
import '../../../../core/errors/business_rule_violation.dart';
import '../../data/customer_photo_repository.dart';
import '../../../../shared/models/customer_photo.dart';
import '../../../../shared/models/customer_photo_status.dart';

/// Selects [photoId] as the customer's profile photo — "one approved
/// photo may be selected," "0 or 1 selected profile photo," "a selected
/// but unapproved photo must not become publicly visible" (enforced
/// structurally: throws [CustomerPhotoNotApprovedViolation] rather than
/// allowing it). Every other photo of the same customer is deselected
/// first, so the 0-or-1 invariant holds without a separate cleanup step.
///
/// P.4.2B2 (2026-08-19): [requestingCustomerId] is now required and
/// checked against [CustomerPhoto.customerId] — this closes the exact
/// ownership gap this class's own doc comment used to name outright
/// ("No authorization gate — self-service, the customer's own choice").
/// **This check is defense-in-depth only, not the security boundary**:
/// this whole class operates against an in-memory repository with no
/// backend behind it. The real, server-authoritative boundary is the
/// `selectCustomerProfilePhoto` Cloud Function
/// (`functions/src/selectCustomerProfilePhoto.ts`), which independently
/// re-verifies ownership (and tenant scope) against the real Firestore
/// `customerPhotos` record — this class being correct does not make a
/// client-side call trustworthy on its own.
class SelectCustomerProfilePhoto {
  const SelectCustomerProfilePhoto(
      {required CustomerPhotoRepository repository})
      : _repository = repository;

  final CustomerPhotoRepository _repository;

  Future<CustomerPhoto> call({
    required String photoId,
    required String requestingCustomerId,
  }) async {
    final photo = await _repository.findById(photoId);
    if (photo == null) {
      throw UnknownAdminEntityViolation(
          entityName: 'CustomerPhoto', id: photoId);
    }
    if (photo.customerId != requestingCustomerId) {
      throw const AuthorizationDeniedViolation(
          actionName: 'selectCustomerProfilePhoto');
    }
    if (photo.status != CustomerPhotoStatus.approved) {
      throw CustomerPhotoNotApprovedViolation(photoId: photoId);
    }

    final siblings = await _repository.findByCustomerId(photo.customerId);
    for (final sibling in siblings) {
      if (sibling.id != photoId && sibling.isSelectedAsProfilePhoto) {
        await _repository.save(sibling.copyWith(
          isSelectedAsProfilePhoto: false,
          revision: sibling.revision + 1,
        ));
      }
    }

    final updated = photo.copyWith(
      isSelectedAsProfilePhoto: true,
      revision: photo.revision + 1,
    );
    await _repository.save(updated);
    return updated;
  }
}

File: functions/src/selectCustomerProfilePhoto.ts
import { onCall, HttpsError } from "firebase-functions/v2/https";
import { getFirestore, Timestamp } from "firebase-admin/firestore";
import * as logger from "firebase-functions/logger";
import { shouldEnforceAppCheck } from "./appCheckConfig";

/**
 * `selectCustomerProfilePhoto` — Profile P.4.2B2 (2026-08-19).
 *
 * Closes the ownership gap `select_customer_profile_photo.dart` itself
 * has documented since P.4/P.4.1 audits: "No authorization gate —
 * self-service, the customer's own choice" was true of the in-memory Dart
 * use case, but nothing server-side ever re-enforced it. This callable
 * IS that server-side boundary — the Dart use case's own defense-in-depth
 * ownership check (added alongside this file) is a second line, never
 * the actual security boundary; only this callable, backed by
 * `firestore.rules`' `customerPhotos`/`customerPublicProfiles` `allow
 * write: if false`, is.
 *
 * **Canonical selected-photo source — locked correction from P.4.1's own
 * plan**: `customers/{uid}.profilePicturePath` is never written here (it
 * stays non-client-writable and is not treated as canonical for anything
 * — it is a single GLOBAL field on the customer identity, unsafe as the
 * source of a TENANT-scoped selection). The one canonical, public-safe
 * source is `customerPublicProfiles/{organizationId}_{uid}
 * .selectedProfilePhotoRef` — already P.4.1's own minimal projection,
 * now actually written to for the first time.
 *
 * **No extra internal-pointer field was added to the projection.** The
 * previously-selected photo (if any) is found via an in-transaction query
 * (`customerPhotos` where `customerId`/`organizationId`/
 * `isSelectedAsProfilePhoto==true`) rather than a stored pointer — this
 * codebase already establishes `tx.get(query)` as a safe, existing
 * pattern (`customerPhotoUploadGrants.ts`'s own quota query), and it
 * avoids adding server-only metadata to a document real staff/customers
 * can otherwise read the whole of. Concurrency safety does not depend on
 * that query's own conflict detection — every selection transaction also
 * reads-then-writes the projection document itself, and Firestore
 * automatically retries the whole transaction body (query included) on a
 * conflicting commit, so a losing transaction re-observes the winner's
 * already-updated state on its retry rather than racing against stale
 * data.
 *
 * **Idempotent**: selecting the photo that is already the customer's
 * selected profile photo is a no-op success — no revision bump, no
 * duplicate audit entry.
 */

interface SelectCustomerProfilePhotoResult {
  photoId: string;
  photoRef: string;
  idempotent: boolean;
}

function db() {
  return getFirestore();
}

export const selectCustomerProfilePhoto = onCall(
  { enforceAppCheck: shouldEnforceAppCheck() },
  async (request): Promise<SelectCustomerProfilePhotoResult> => {
    if (!request.auth) {
      throw new HttpsError("unauthenticated", "Sign-in is required.");
    }
    const isRealCustomer = request.auth.token?.firebase?.sign_in_provider === "phone";
    if (!isRealCustomer) {
      throw new HttpsError(
        "permission-denied",
        "Selecting a profile photo requires a real, phone-verified customer identity.",
      );
    }
    const uid = request.auth.uid;

    const photoId = request.data?.photoId;
    if (typeof photoId !== "string" || photoId.length === 0) {
      throw new HttpsError("invalid-argument", "photoId is required.");
    }
    const organizationId = request.data?.organizationId;
    if (typeof organizationId !== "string" || organizationId.length === 0) {
      throw new HttpsError("invalid-argument", "organizationId is required.");
    }

    // Never trust the client's organizationId as authorization truth on
    // its own — independently verify real tenant membership, same
    // mechanism `requestCustomerPhotoUploadGrant` already uses.
    const tenantCustomerDoc = await db()
      .collection("tenantCustomers")
      .doc(`${organizationId}_${uid}`)
      .get();
    if (!tenantCustomerDoc.exists) {
      throw new HttpsError("permission-denied", "You are not a customer of this organization.");
    }

    const photoRef = db().collection("customerPhotos").doc(photoId);
    const projectionRef = db().collection("customerPublicProfiles").doc(`${organizationId}_${uid}`);

    const result = await db().runTransaction(
      async (tx): Promise<SelectCustomerProfilePhotoResult> => {
        // Reads before writes: target photo, the projection (may not
        // exist yet), and every OTHER currently-selected photo of this
        // same customer within this same tenant.
        const [photoSnap, projectionSnap, currentlySelectedSnap] = await Promise.all([
          tx.get(photoRef),
          tx.get(projectionRef),
          tx.get(
            db()
              .collection("customerPhotos")
              .where("customerId", "==", uid)
              .where("organizationId", "==", organizationId)
              .where("isSelectedAsProfilePhoto", "==", true),
          ),
        ]);

        if (!photoSnap.exists) {
          throw new HttpsError("not-found", "Customer photo not found.");
        }
        const photo = photoSnap.data()!;

        // Ownership: caller may select ONLY their own photo.
        if (photo.customerId !== uid) {
          throw new HttpsError("permission-denied", "You may only select your own photo.");
        }
        // Tenant scope: the photo must belong to the requested/current
        // tenant — a cross-tenant photoId can never be selected here
        // even if (implausibly) it also happened to belong to this uid.
        if (photo.organizationId !== organizationId) {
          throw new HttpsError("permission-denied", "This photo does not belong to the requested organization.");
        }
        if (photo.status !== "approved") {
          throw new HttpsError(
            "failed-precondition",
            "Only an approved photo may be selected as the profile photo.",
          );
        }

        const projectionData = projectionSnap.exists ? projectionSnap.data()! : null;
        if (projectionData?.selectedProfilePhotoRef === photo.photoRef) {
          // Already selected — idempotent success, no mutation.
          return { photoId, photoRef: photo.photoRef as string, idempotent: true };
        }

        const now = Timestamp.now();

        for (const doc of currentlySelectedSnap.docs) {
          if (doc.id === photoId) continue; // defensive — shouldn't appear given the check above.
          const sibling = doc.data();
          tx.update(doc.ref, {
            isSelectedAsProfilePhoto: false,
            revision: (sibling.revision as number) + 1,
          });
        }

        tx.update(photoRef, {
          isSelectedAsProfilePhoto: true,
          revision: (photo.revision as number) + 1,
        });

        tx.set(
          projectionRef,
          {
            uid,
            organizationId,
            selectedProfilePhotoRef: photo.photoRef,
            updatedAt: now,
          },
          { merge: true },
        );

        tx.set(db().collection("auditEvents").doc(), {
          organizationId,
          uid,
          actorId: uid,
          photoId,
          previousSelectedPhotoId: currentlySelectedSnap.docs.find((d) => d.id !== photoId)?.id ?? null,
          type: "customerPhoto.selected",
          timestamp: now,
        });

        return { photoId, photoRef: photo.photoRef as string, idempotent: false };
      },
    );

    logger.info("[selectCustomerProfilePhoto] processed", {
      uid,
      organizationId,
      photoId,
      idempotent: result.idempotent,
    });

    return result;
  },
);
